File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/menstruation_period_record_mappers.dart
import 'package:health_connector_core/health_connector_core.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mappers.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_platform_api.g.dart';
import 'package:meta/meta.dart';

@internal
extension MenstruationPeriodRecordToDto on MenstruationPeriodRecord {
  MenstruationPeriodRecordDto toDto() {
    return MenstruationPeriodRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDto(),
      samples: samples.map((e) => e.toDto()).toList(),
    );
  }
}

@internal
extension MenstruationPeriodRecordDtoToDomain on MenstruationPeriodRecordDto {
  MenstruationPeriodRecord toDomain() {
    return MenstruationPeriodRecord(
      id: id?.toHealthRecordId() ?? HealthRecordId.none,
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime, isUtc: true),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDomain(),
      samples: samples
          .whereType<MenstruationFlowRecordDto>()
          .map((e) => e.toDomain())
          .toList(),
    );
  }
}

@internal
extension MenstruationFlowRecordToDto on MenstruationFlowRecord {
  MenstruationFlowRecordDto toDto() {
    return MenstruationFlowRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      flow: _toMenstrualFlowDto(flow),
      metadata: metadata?.toDto(),
      zoneOffsetSeconds: zoneOffsetSeconds,
    );
  }
}

@internal
extension MenstruationFlowRecordDtoToDomain on MenstruationFlowRecordDto {
  MenstruationFlowRecord toDomain() {
    return MenstruationFlowRecord(
      id: id?.toHealthRecordId() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      flow: _toMenstrualFlow(flow),
      metadata: metadata?.toDomain(),
      zoneOffsetSeconds: zoneOffsetSeconds,
    );
  }
}

MenstrualFlowDto? _toMenstrualFlowDto(MenstrualFlow? flow) {
  if (flow == null) {
    return null;
  }
  switch (flow) {
    case MenstrualFlow.light:
      return MenstrualFlowDto.light;
    case MenstrualFlow.medium:
      return MenstrualFlowDto.medium;
    case MenstrualFlow.heavy:
      return MenstrualFlowDto.heavy;
    case MenstrualFlow.unknown:
      return MenstrualFlowDto.unknown;
  }
}

MenstrualFlow _toMenstrualFlow(MenstrualFlowDto? dto) {
  switch (dto) {
    case MenstrualFlowDto.light:
      return MenstrualFlow.light;
    case MenstrualFlowDto.medium:
      return MenstrualFlow.medium;
    case MenstrualFlowDto.heavy:
      return MenstrualFlow.heavy;
    case MenstrualFlowDto.unknown:
    case null:
      return MenstrualFlow.unknown;
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/MenstruationPeriodRecordMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.MenstruationFlowRecord
import androidx.health.connect.client.records.MenstruationPeriodRecord
import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.MenstrualFlowDto
import com.phamtunglam.health_connector_hc_android.pigeon.MenstruationFlowRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MenstruationPeriodRecordDto
import java.time.Instant
import java.time.ZoneOffset

internal fun MenstrualFlowDto.toHealthConnect(): Int = when (this) {
    MenstrualFlowDto.UNKNOWN -> MenstruationFlowRecord.FLOW_UNKNOWN
    MenstrualFlowDto.LIGHT -> MenstruationFlowRecord.FLOW_LIGHT
    MenstrualFlowDto.MEDIUM -> MenstruationFlowRecord.FLOW_MEDIUM
    MenstrualFlowDto.HEAVY -> MenstruationFlowRecord.FLOW_HEAVY
}

internal fun Int.toMenstrualFlowDto(): MenstrualFlowDto = when (this) {
    MenstruationFlowRecord.FLOW_UNKNOWN -> MenstrualFlowDto.UNKNOWN
    MenstruationFlowRecord.FLOW_LIGHT -> MenstrualFlowDto.LIGHT
    MenstruationFlowRecord.FLOW_MEDIUM -> MenstrualFlowDto.MEDIUM
    MenstruationFlowRecord.FLOW_HEAVY -> MenstrualFlowDto.HEAVY
    else -> MenstrualFlowDto.UNKNOWN
}

internal fun MenstruationFlowRecord.toDto(): MenstruationFlowRecordDto = MenstruationFlowRecordDto(
    id = metadata.id,
    time = time.toEpochMilli(),
    flow = flow.toMenstrualFlowDto(),
    metadata = metadata.toDto(),
    zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
)

internal fun MenstruationFlowRecordDto.toHealthConnect(): MenstruationFlowRecord =
    MenstruationFlowRecord(
        time = Instant.ofEpochMilli(time),
        flow = flow?.toHealthConnect() ?: MenstruationFlowRecord.FLOW_UNKNOWN,
        metadata = metadata?.toHealthConnect() ?: Metadata.manualEntry(),
        zoneOffset = zoneOffsetSeconds?.let {
            ZoneOffset.ofTotalSeconds(it.toInt())
        } ?: ZoneOffset.UTC,
    )

/**
 * Converts a Health Connect [MenstruationPeriodRecord] to a [MenstruationPeriodRecordDto].
 * Note: Samples are NOT populated here as they require a separate query.
 */
internal fun MenstruationPeriodRecord.toDto(): MenstruationPeriodRecordDto =
    MenstruationPeriodRecordDto(
        id = metadata.id,
        metadata = metadata.toDto(),
        startTime = startTime.toEpochMilli(),
        endTime = endTime.toEpochMilli(),
        startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
        endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
        samples = emptyList(), // Samples must be populated by the Handler
    )

/**
 * Converts a [MenstruationPeriodRecordDto] to a Health Connect [MenstruationPeriodRecord].
 * Note: Samples are ignored as they are stored as separate records.
 */
internal fun MenstruationPeriodRecordDto.toHealthConnect(): MenstruationPeriodRecord =
    MenstruationPeriodRecord(
        startTime = Instant.ofEpochMilli(startTime),
        endTime = Instant.ofEpochMilli(endTime),
        startZoneOffset = startZoneOffsetSeconds?.let {
            ZoneOffset.ofTotalSeconds(it.toInt())
        } ?: ZoneOffset.UTC,
        endZoneOffset = endZoneOffsetSeconds?.let {
            ZoneOffset.ofTotalSeconds(it.toInt())
        } ?: ZoneOffset.UTC,
        metadata = metadata.toHealthConnect(),
    )
